In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.svm import SVC
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report, precision_score, recall_score, f1_score
from sklearn.metrics import roc_auc_score
from sklearn.preprocessing import PolynomialFeatures
%matplotlib inline

In [2]:
data_train = pd.read_csv('data_train_ros.csv')
data_test = pd.read_csv('data_test.csv')

In [3]:
x_train = data_train[['distinct_ref_no_3month', 'distinct_ref_no_month', 'trans_group_3month_1', 'diff_amount_3month',
          'diff_amount_week', 'diff_amount_month', 'trans_group_month_1.0', 'amount_3month', 'distinct_ref_no_week',
          'amount_month', 'max_amount_3month']]
y_train = data_train['Is_churn']
x_test = data_test[['distinct_ref_no_3month', 'distinct_ref_no_month', 'trans_group_3month_1', 'diff_amount_3month',
          'diff_amount_week', 'diff_amount_month', 'trans_group_month_1.0', 'amount_3month', 'distinct_ref_no_week',
          'amount_month', 'max_amount_3month']]
y_test = data_test['Is_churn']

In [5]:
svc = SVC(kernel='rbf')
svc.fit(x_train, y_train)
y_pred = svc.predict(x_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.98      0.74      0.84     10459
           1       0.11      0.67      0.19       499

    accuracy                           0.74     10958
   macro avg       0.55      0.71      0.52     10958
weighted avg       0.94      0.74      0.81     10958



In [6]:
roc_auc_score(y_test, y_pred)


0.7075022978359434

In [7]:
y_pred = svc.predict(x_train)
print(classification_report(y_train, y_pred))

              precision    recall  f1-score   support

           0       0.73      0.74      0.73     41835
           1       0.73      0.72      0.73     41835

    accuracy                           0.73     83670
   macro avg       0.73      0.73      0.73     83670
weighted avg       0.73      0.73      0.73     83670



In [9]:
roc_auc_score(y_train, y_pred)


0.7301780805545595

## Conclusion

Based on these above information, we can see that the K_Nearest Neighbors model is
the most efficient model with Accuracy reaching 94%, AUC reaching 66.5% and F1-score reaching 37%. Do
Therefore, we have selected K-Nearest Neighbors as the Churn Customer forecasting model.